### PR TITLE
DAOS-7707 tests: Adding method to adjust pool size to the configuration.

### PR DESCRIPTION
**Pull request by phender**
Adding two new TestWithServers methods to allow tests to create pools
that utilize a percentage of the available SCM and/or NVMe space
configured by the active servers.  These methods also provide the option
of reducing the number of targets per engine in the server configuration
if the current server configuration cannot support the requested pool
size.

Test-tag: pool_create_tests dmg_system_reformat

Signed-off-by: Phillip Henderson <phillip.henderson@intel.com>

### CODE REVIEW

**Comment by phender** [comment]
When testing these changes on the OFED cluster (https://github.com/daos-stack/daos/pull/6003) I did run into https://jira.hpdd.intel.com/browse/DAOS-7795

**Comment by brianjmurrell** [comment]
@phender Could you merge master again, (or rebase) on this at least.  If you wanted to squash and force push it (since you have not asked for reviews yet) I wouldn't object to that either.  :-)

**Comment by phender** [comment]
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5940/17/ completed successfully with the tests using the new pool sizing feature.  Running now with all pr and daily tests to verify there are no adverse affects.

**Review by shimizukko — APPROVED** [review]

**Comment by phender** [comment]
Failures unrelated to the PR changes seen in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5940/18/:

- ./container/autotest.py:ContainerAutotestTest.test_container_autotest: avocado test timeout: https://jira.hpdd.intel.com/browse/DAOS-7946
- ./osa/osa_offline_reintegration.py:OSAOfflineReintegration.test_osa_offline_reintegration_with_rf: daos container check SIGSEGV: https://jira.hpdd.intel.com/browse/DAOS-7948

Still waiting for HW Large results.

**Review by shimizukko — APPROVED** [review]

**Review by shimizukko — DISMISSED** [review]

File: src/tests/ftest/nvme/nvme_object.yaml [hunk 1]
```diff
@@ -29,20 +29,24 @@ agent_config:
...
-      - 20000000000
-      - 100000000000
-      - 800000000000
-  createsvc:
-    svcn: 1
+pool_1:
   control_method: dmg
+  name: daos_server
+  scm_size: 4000000000
+  nvme_size: 20000000000
```
Comment (brianjmurrell) [inline src/tests/ftest/nvme/nvme_object.yaml:36]: I wish we'd use human-readable values in these files.  I have a PR that moves all of these values to human-readable values in all yaml files.  Most assuredly bit-rotten by now.  :frowning:

**Review by brianjmurrell — COMMENTED** [review]
While this LGTM, I should probably preserve my +1 slot for somebody with more contextual knowledge of the code being updated here.

**Review by shimizukko — APPROVED** [review]

File: src/tests/ftest/nvme/nvme_object.yaml (hunk 1 above)
Comment (ravalsam) [reply src/tests/ftest/nvme/nvme_object.yaml:36]: I don't think human-readable values were supported when some of this test was developed. That is the reason many of the test codes still use the plain size including rebuild/container/pool test folders.

Comment (brianjmurrell) [reply src/tests/ftest/nvme/nvme_object.yaml:36]: Yes, completely understood.  But that doesn't apply any more and IMHO, human readable values should be the norm going forward.

**Review by mjean308 — APPROVED** [review]

File: src/tests/ftest/pool/create_test.yaml [hunk 2]
```diff
@@ -23,7 +23,16 @@ server_config:
       scm_class: dcpm
       scm_list: ["/dev/pmem0"]
       scm_mount: /mnt/daos0
-pool:
+pool_1:
   name: daos_server
   control_method: dmg
-  quantity: 200
+  scm_size: 90%
+  svcn: 1
```
Comment (ravalsam) [inline src/tests/ftest/pool/create_test.yaml:30]: This should be the default and no need to pass it?

File: src/tests/ftest/control/dmg_system_reformat.py [hunk 3]
```diff
@@ -51,19 +49,27 @@ def test_dmg_system_reformat(self):
...
         # Uncomment below after DAOS-5979 is resolved
         # self.assertTrue(
         #     self.server_managers[-1].dmg.set_config_value("hostlist", None))
 
         self.log.info("Perform dmg storage format on all system ranks:")
-        self.get_dmg_command().storage_format(reformat=True)
+        self.get_dmg_command().storage_format(force=True)
         if self.get_dmg_command().result.exit_status != 0:
-            self.fail("Issues performing storage format --reformat: {}".format(
+            self.fail("Issues performing storage format --force: {}".format(
```
Comment (ravalsam) [inline src/tests/ftest/control/dmg_system_reformat.py:72]: The test is intended to verify reformat function so any reason to change to --force?

File: src/tests/ftest/nvme/nvme_object.yaml (hunk 1 above)
Comment (ravalsam) [reply src/tests/ftest/nvme/nvme_object.yaml:36]: I agree if we are changing anyway better to use a human-readable format.

File: src/tests/ftest/pool/create_capacity_test.yaml [hunk 4]
```diff
@@ -24,4 +24,7 @@ server_config:
 pool:
   name: daos_server
   control_method: dmg
+  scm_size: 60%
+  nvme_size: 60%
```
Comment (ravalsam) [inline src/tests/ftest/pool/create_capacity_test.yaml:28]: So this will create 200 pools out of 60% of the capacity?

**Review by ravalsam — CHANGES_REQUESTED** [review]

File: src/tests/ftest/control/dmg_system_reformat.py (hunk 3 above)
Comment (phender) [reply src/tests/ftest/control/dmg_system_reformat.py:72]: When attempting to verify my changes did not adversely affect this test I was attempting to get the test to pass with similar changes to what @mjmac is doing in https://github.com/daos-stack/daos/pull/5203/files.  I was able to verify my pool creation changes but the test still did not pass with the other changes.  Ultimately I restored the `@skipForTicket("DAOS-6004")` and the test is still skipped.

File: src/tests/ftest/nvme/nvme_object.py [hunk 5]
```diff
@@ -188,19 +176,19 @@ def test_nvme_object_multiple_pools(self):
             Verify the objects are being created and the data is not
             corrupted.
 
-        :avocado: tags=all,full_regression,hw,large,nvme_object_multiple_pools
-        :avocado: tags=nvme_object
+        :avocado: tags=all,full_regression
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/nvme/nvme_object.py:179]: Ah, no.  It's a weekly test and I did not include the tag.  I did test this manually and it worked:
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuite errors="0" failures="0" name="job-2021-06-25T16.51-ba7d4c3" skipped="0" tests="2" time="1123.972" timestamp="2021-06-25T17:10:20.684640">
        <testcase classname="NvmeObject" name="1-./nvme/nvme_object.py:NvmeObject.test_nvme_object_single_pool;transport_config-container-transport_config-hosts-pool_1-pool_2-pool_3-servers-transport_config-timeouts-abdd" time="42.139"/>
        <testcase classname="NvmeObject" name="2-./nvme/nvme_object.py:NvmeObject.test_nvme_object_multiple_pools;transport_config-container-transport_config-hosts-pool_1-pool_2-pool_3-servers-transport_config-timeouts-abdd" time="1081.833"/>
</testsuite>
```

File: src/tests/ftest/pool/create_test.yaml (hunk 2 above)
Comment (phender) [reply src/tests/ftest/pool/create_test.yaml:30]: The default for the `svcn` TestPool parameter is None.  This setting ensures that `--nsvc=1` is included in the `dmg pool create` command.

File: src/tests/ftest/pool/create_capacity_test.yaml (hunk 4 above)
Comment (phender) [reply src/tests/ftest/pool/create_capacity_test.yaml:28]: Yes, the combination of `quantity: 200` with `scm_size: 60%` and `nvme_size: 60%` will yield a single pool with a `dmg pool create` command with `--scm_size=~(<SCM_per_engine> * 0.6) / 200` and `nvme_size=~(<NVMe_per_engine> * 0.6) / 200`.  Then if the code calls `self.add_pool_qty(200)` then 200x pools will be created with these sizes:
```
2021-06-25 23:04:08,086 server_utils     L0820 INFO | ----------------------------------------------------------------------------------------------------
2021-06-25 23:04:08,086 server_utils     L0824 INFO | Autosizing TestPool parameters ending with a "%" for 200 pools:
2021-06-25 23:04:08,086 server_utils     L0827 INFO |   - size      : None (False)
2021-06-25 23:04:08,086 server_utils     L0827 INFO |   - scm_size  : 60% (True)
2021-06-25 23:04:08,086 server_utils     L0827 INFO |   - nvme_size : 60% (True)
2021-06-25 23:04:08,086 server_utils_bas L0502 INFO | Detected device capacities:
2021-06-25 23:04:08,086 server_utils_bas L0509 INFO |   NVME for 0000:86:00.0 : ['750156374016 (698.64GiB) (750.16GB)']
2021-06-25 23:04:08,086 server_utils_bas L0509 INFO |   NVME for 0000:87:00.0 : ['750156374016 (698.64GiB) (750.16GB)']
2021-06-25 23:04:08,086 server_utils_bas L0509 INFO |   SCM  for pmem0 : ['3183575302144 (2.9TiB) (3.18TB)']
2021-06-25 23:04:08,086 server_utils_bas L0509 INFO |   SCM  for pmem1 : ['3183575302144 (2.9TiB) (3.18TB)']
2021-06-25 23:04:08,086 server_utils_bas L0536 INFO | Detected engine capacities:
2021-06-25 23:04:08,087 server_utils_bas L0540 INFO |   NVME : ['1500312748032 (1.36TiB) (1.5TB)']
2021-06-25 23:04:08,087 server_utils_bas L0540 INFO |   SCM  : ['3183575302144 (2.9TiB) (3.18TB)']
2021-06-25 23:04:08,087 server_utils     L0755 INFO | Largest storage size available per engine:
2021-06-25 23:04:08,087 server_utils     L0760 INFO |   NVME:  1500312748032 (1.36TiB) (1.5TB)
2021-06-25 23:04:08,087 server_utils     L0760 INFO |   SCM :  3183575302144 (2.9TiB) (3.18TB)
2021-06-25 23:04:08,087 server_utils     L0851 INFO | Largest storage size available for 7 engines with a 6.00% scm_ratio:
2021-06-25 23:04:08,087 server_utils     L0854 INFO |   - NVME     : 10502189236224 (9.55TiB) (10.5TB)
2021-06-25 23:04:08,087 server_utils     L0857 INFO |   - SCM      : 630131354173.44 (586.86GiB) (630.13GB)
2021-06-25 23:04:08,087 server_utils     L0861 INFO |   - COMBINED : 11132320590397.44 (10.12TiB) (11.13TB)
2021-06-25 23:04:08,087 server_utils     L0869 INFO | Adjusted pool sizes for 200 pools:
2021-06-25 23:04:08,088 server_utils     L0882 INFO |   - scm_size  : SCM  storage adjusted by 60.00%: 9550725906.432 (8.89GiB) (9.55GB)
2021-06-25 23:04:08,088 server_utils     L0882 INFO |   - nvme_size : NVMe storage adjusted by 60.00%: 4500938244.096 (4.19GiB) (4.5GB)
2021-06-25 23:04:08,088 server_utils     L0889 INFO | Increment sizes per target:
2021-06-25 23:04:08,088 server_utils     L0892 INFO |   - scm_size  : 16777216 (16.0MiB) (16.78MB)
2021-06-25 23:04:08,088 server_utils     L0892 INFO |   - nvme_size : 1073741824 (1024.0MiB) (1.07GB)
2021-06-25 23:04:08,088 server_utils     L0895 INFO | Pool sizes adjusted to fit by increment sizes:
2021-06-25 23:04:08,088 server_utils     L0903 INFO |   - scm_size  : 569 * 16777216 = 9546235904 (8.89GiB) (9.55GB)
2021-06-25 23:04:08,089 server_utils     L0903 INFO |   - nvme_size : 4 * 1073741824 = 4294967296 (4.0GiB) (4.29GB)
2021-06-25 23:04:08,089 server_utils     L0931 INFO | ----------------------------------------------------------------------------------------------------
2021-06-25 23:04:08,089 command_utils_ba L0116 DEBUG| Updated param scm_ratio => None
2021-06-25 23:04:08,089 command_utils_ba L0116 DEBUG| Updated param scm_size => 8.89GiB
2021-06-25 23:04:08,089 command_utils_ba L0116 DEBUG| Updated param nvme_size => 4.0GiB
2021-06-25 23:04:08,110 test_utils_pool  L0144 INFO | Creating a pool
2021-06-25 23:04:08,111 general_utils    L0260 INFO | Command environment vars:
  None
2021-06-25 23:04:08,111 process          L0626 INFO | Running '/usr/bin/dmg -o /etc/daos/daos_control.yml -d -j pool create --group=jenkins --nsvc=1 --nvme-size=4.0GiB --scm-size=8.89GiB --user=jenkins'
```

File: src/tests/ftest/nvme/nvme_object.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/nvme/nvme_object.py:179]: Thanks for confirming.

**Review by ravalsam — APPROVED** [review]

File: src/tests/ftest/util/dmg_utils.py [hunk 6]
```diff
@@ -823,7 +831,7 @@ def system_query(self, ranks=None, verbose=True):
             ("system", "query"), ranks=ranks, verbose=verbose)
 
     def system_leader_query(self):
-        """Query system to obtain the MS leader and replica information.
+        """Erase the system to obtain the MS leader and replica information.
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/dmg_utils.py]: That was an old typo I fixed in https://github.com/daos-stack/daos/pull/5940/commits/c94fc22a5d1d0b32961a967338b4d0dd92f11e69

**Comment by brianjmurrell** [comment]
Is this ready to land yet?

**Comment by phender** [comment]
> Is this ready to land yet?

The last full run over the weekend (https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5940/26/) was essentially successful - the failures were known issues.  I did have to merge with master to resolve a simple conflict and that run terminated after https://jira.hpdd.intel.com/browse/DAOS-7967, so there were no functional HW results.  Waiting to hear if between this and the https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6003 testing if more testing is required to land these changes.

**Pull request merged**